src/compile.c

```diff
@@ -86,7 +86,10 @@ u8 *machine_compile(machine_t *m, str_t source) {
 
             elf64_sym_t *sym = (elf64_sym_t *)(elfbuf + symtab_shdr->sh_offset + rel->r_sym * sizeof(elf64_sym_t));
             u32 *loc = (u32 *)(text_addr + rel->r_offset);
-            *loc = (u32)((i64)sym->st_value + rel->r_addend - (i64)rel->r_offset);
+            *loc = (u32)((rodata_addr - text_addr) + sym->st_value + rel->r_addend - rel->r_offset);
+            // ensure relocation correctly
```

Comment:
I suggest removing this comment, it does not provide additional information for the reader. But I think it would be nice to define three variables: S, A, and P, using them for the calculation, and giving comments on them.